File: lib/message/cslib/src/cslib_wrap.f90
! ISO_C_binding wrapper on CSlib C interface

module cslib_wrap

interface
  subroutine cslib_open_fortran(csflag,mode,str,pcomm,ptr) bind(c)
    use iso_c_binding
    integer(c_int), value :: csflag
    character(c_char) :: mode(*),str(*)
    type(c_ptr), value :: pcomm
    type(c_ptr) :: ptr
  end subroutine cslib_open_fortran

  subroutine cslib_open_fortran_mpi_one(csflag,mode,pboth,pcomm,ptr) bind(c)
    use iso_c_binding
    integer(c_int), value :: csflag
    character(c_char) :: mode(*)
    type(c_ptr), value :: pboth,pcomm
    type(c_ptr) :: ptr
  end subroutine cslib_open_fortran_mpi_one

  subroutine cslib_close(ptr) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
  end subroutine cslib_close

  subroutine cslib_send(ptr,msgID,nfield) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: msgID,nfield
  end subroutine cslib_send

  subroutine cslib_pack_int(ptr,id,value) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
    integer(c_int), value :: value
  end subroutine cslib_pack_int

  subroutine cslib_pack_int64(ptr,id,value) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
    integer(c_int64_t), value :: value
  end subroutine cslib_pack_int64

  subroutine cslib_pack_float(ptr,id,value) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
    real(c_float), value :: value
  end subroutine cslib_pack_float

  subroutine cslib_pack_double(ptr,id,value) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
    real(c_double), value :: value
  end subroutine cslib_pack_double

  subroutine cslib_pack_string(ptr,id,value) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
    character(c_char) :: value(*)
  end subroutine cslib_pack_string

  subroutine cslib_pack(ptr,id,ftype,flen,data) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id,ftype,flen
    type(c_ptr), value :: data
  end subroutine cslib_pack

  subroutine cslib_pack_parallel(ptr,id,ftype,nlocal,ids,nper,data) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id,ftype,nlocal,nper
    type(c_ptr), value :: ids,data
  end subroutine cslib_pack_parallel

  function cslib_recv(ptr,nfield,fieldID,fieldtype,fieldlen) bind(c)
    use iso_c_binding
    integer(c_int) :: cslib_recv
    type(c_ptr), value :: ptr
    integer(c_int) :: nfield
    type(c_ptr) :: fieldID,fieldtype,fieldlen
  end function cslib_recv

  function cslib_unpack_int(ptr,id) bind(c)
    use iso_c_binding
    integer(c_int) :: cslib_unpack_int
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack_int

  function cslib_unpack_int64(ptr,id) bind(c)
    use iso_c_binding
    integer(c_int64_t) :: cslib_unpack_int64
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack_int64

  function cslib_unpack_float(ptr,id) bind(c)
    use iso_c_binding
    real(c_float) :: cslib_unpack_float
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack_float

  function cslib_unpack_double(ptr,id) bind(c)
    use iso_c_binding
    real(c_double) :: cslib_unpack_double
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack_double

  function cslib_unpack_string(ptr,id) bind(c)
    use iso_c_binding
    type(c_ptr) :: cslib_unpack_string
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack_string

  function cslib_unpack(ptr,id) bind(c)
    use iso_c_binding
    type(c_ptr) :: cslib_unpack
    type(c_ptr), value :: ptr
    integer(c_int), value :: id
  end function cslib_unpack

  subroutine cslib_unpack_parallel(ptr,id,nlocal,ids,nper,data) bind(c)
    use iso_c_binding
    type(c_ptr), value :: ptr
    integer(c_int), value :: id,nlocal,nper
    type(c_ptr), value :: ids,data
  end subroutine cslib_unpack_parallel

  function cslib_extract(ptr,flag) bind(c)
    use iso_c_binding
    integer(c_int) :: cslib_extract
    type(c_ptr), value :: ptr
    integer(c_int), value :: flag
  end function cslib_extract
end interface

end module cslib_wrap

File: lib/message/cslib/src/cslib_wrap.cpp
/* ----------------------------------------------------------------------
   CSlib - Client/server library for code coupling
   https://cslib.sandia.gov/, Sandia National Laboratories
   Steve Plimpton, sjplimp@sandia.gov

   Copyright 2018 National Technology & Engineering Solutions of
   Sandia, LLC (NTESS). Under the terms of Contract DE-NA0003525 with
   NTESS, the U.S. Government retains certain rights in this software.
   This software is distributed under the modified Berkeley Software
   Distribution (BSD) License.

   See the README file in the top-level CSlib directory.
------------------------------------------------------------------------- */

// C style library interface to CSlib class

#ifdef MPI_YES
#include <mpi.h>
#else
#include <mpi_dummy.h>
#endif
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cslib_wrap.h"
#include "cslib.h"

using namespace CSLIB_NS;

// ----------------------------------------------------------------------

void cslib_open(int csflag, const char *mode, const void *ptr,
                const void *pcomm, void **csptr)
{
  CSlib *cs = new CSlib(csflag,mode,ptr,pcomm);
  *csptr = (void *) cs;
}

// ----------------------------------------------------------------------

void cslib_open_fortran(int csflag, const char *mode, const char *str,
                        const void *pcomm, void **csptr)
{
  MPI_Comm ccomm;
  void *pccomm = nullptr;

  if (pcomm) {
    MPI_Fint *fcomm = (MPI_Fint *) pcomm;
    ccomm = MPI_Comm_f2c(*fcomm);
    pccomm = &ccomm;
  }

  CSlib *cs = new CSlib(csflag,mode,str,pccomm);
  *csptr = (void *) cs;
}

// ----------------------------------------------------------------------

void cslib_open_fortran_mpi_one(int csflag, const char *mode,
                                const void *pboth, const void *pcomm,
                                void **csptr)
{
  MPI_Comm ccomm,cboth;
  void *pccomm,*pcboth;

  MPI_Fint *fcomm = (MPI_Fint *) pcomm;
  ccomm = MPI_Comm_f2c(*fcomm);
  pccomm = &ccomm;

  MPI_Fint *fboth = (MPI_Fint *) pboth;
  cboth = MPI_Comm_f2c(*fboth);
  pcboth = &cboth;

  CSlib *cs = new CSlib(csflag,mode,pcboth,pccomm);
  *csptr = (void *) cs;
}

// ----------------------------------------------------------------------

void cslib_close(void *ptr)
{
  CSlib *cs = (CSlib *) ptr;
  delete cs;
}

// ----------------------------------------------------------------------

void cslib_send(void *ptr, int msgID, int nfield)
{
  CSlib *cs = (CSlib *) ptr;
  cs->send(msgID,nfield);
}

// ----------------------------------------------------------------------

void cslib_pack_int(void *ptr, int id, int value)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_int(id,value);
}

// ----------------------------------------------------------------------

void cslib_pack_int64(void *ptr, int id, int64_t value)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_int64(id,value);
}

// ----------------------------------------------------------------------

void cslib_pack_float(void *ptr, int id, float value)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_float(id,value);
}

// ----------------------------------------------------------------------

void cslib_pack_double(void *ptr, int id, double value)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_double(id,value);
}

// ----------------------------------------------------------------------

void cslib_pack_string(void *ptr, int id, char *value)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_string(id,value);
}

// ----------------------------------------------------------------------

void cslib_pack(void *ptr, int id, int ftype, int flen, void *data)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack(id,ftype,flen,data);
}

// ----------------------------------------------------------------------

void cslib_pack_parallel(void *ptr, int id, int ftype,
			 int nlocal, int *ids, int nper, void *data)
{
  CSlib *cs = (CSlib *) ptr;
  cs->pack_parallel(id,ftype,nlocal,ids,nper,data);
}

// ----------------------------------------------------------------------

int cslib_recv(void *ptr, int *nfield_caller,
	       int **fieldID_caller, int **fieldtype_caller,
	       int **fieldlen_caller)
{
  CSlib *cs = (CSlib *) ptr;

  int nfield;
  int *fieldID,*fieldtype,*fieldlen;
  int msgID = cs->recv(nfield,fieldID,fieldtype,fieldlen);

  *nfield_caller = nfield;
  *fieldID_caller = fieldID;
  *fieldtype_caller = fieldtype;
  *fieldlen_caller = fieldlen;

  return msgID;
}

// ----------------------------------------------------------------------

int cslib_unpack_int(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack_int(id);
}
// ----------------------------------------------------------------------

int64_t cslib_unpack_int64(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack_int64(id);
}

// ----------------------------------------------------------------------

float cslib_unpack_float(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack_float(id);
}

// ----------------------------------------------------------------------

double cslib_unpack_double(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack_double(id);
}

// ----------------------------------------------------------------------

char *cslib_unpack_string(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack_string(id);
}

// ----------------------------------------------------------------------

void *cslib_unpack(void *ptr, int id)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->unpack(id);
}

// ----------------------------------------------------------------------

void cslib_unpack_data(void *ptr, int id, void *data)
{
  CSlib *cs = (CSlib *) ptr;
  cs->unpack(id,data);
}

// ----------------------------------------------------------------------

void cslib_unpack_parallel(void *ptr, int id, int nlocal, int *ids,
			   int nper, void *data)
{
  CSlib *cs = (CSlib *) ptr;
  cs->unpack_parallel(id,nlocal,ids,nper,data);
}

// ----------------------------------------------------------------------

int cslib_extract(void *ptr, int flag)
{
  CSlib *cs = (CSlib *) ptr;
  return cs->extract(flag);
}
